### PR TITLE
revamped c bindings 

### PR DESCRIPTION
**Pull request by koltenpearson**
I am trying to integrate pocketpy with a c program, and I could not figure out how to make the single file header's extern c section work for what I needed to do.

So I am proposing a different way to manage c bindings. I made a pocketpy_c.h and pocketpy_c.cpp where the bindings can be placed. A c project can compile those as a library with the c++ compiler and then link it to a c program that includes the pocketpy_c.h to be able to use them.

Unfortunately c projects lose the single file header setup, but they were already going to incur enough complexity managing the mixed c/c++ build that adding two additional files for them is not really a dealbreaker. 

The current state of the code can compile and run a hello world example using c. I wanted to get approval for the new setup before I got any farther. For my purposes I would be willing to improve the pull request at least until it was able to launch a vm, bind c functions to python, call python functions from c, and evaluate/consume python code from a string.

### CODE REVIEW

**Comment by blueloveTH** [comment]
Well. pocketpy does not support C bindings via source integration now.
Some `extern "C"` functions are designed to work with dynamic library (.dll .so). In this case, one complies the whole header into a library and the exported C functions can be used via a foreign language. It does not require a split header file.

If you are doing source integration, it is required to use C++ API instead of C API.

**Comment by blueloveTH** [comment]
Another point is, pocketpy highly depends on C++ template/overload to achieve core features of the interpreter, such as customizing gc process for `struct`. It is quite impossible to write a bunch of new APIs with template usage removed.

**Comment by koltenpearson** [comment]
I guess the problem I was having with the `extern "C"` functions that pocketpy.h provides is that pocketpy.h can not be included by a c source file and successfully compiled. It would need to wrap most of its implementation in `#ifdef __cplusplus` for that to work, and then we would also need to separate out the implementation of the extern functions from their declaration since c is not able to handle those the same way c++ can.

The way I have set things up this PR does what you said (you compile the c++ code separately with a c++ compiler and then link it to the c code). But I think I can push the C API little farther and get it working for more functionality than just passing text across, even with those constraints. 

My biggest worry is figuring out how to successfully wrap an ArgsView object, since they are pass by value. But if I can get that working we should be fine for the functionality I outlined above. 

But if you would rather I can do all of this downstream, since the bindings I am writing won't require any changes to the c++ code itself. So if you not interested in expanding the C API functionality in this way then I guess just close this PR and I will worry about it there.

**Comment by koltenpearson** [comment]
Okay I finished the basic proof of concept for a lua inspired api

In the end I did have to modify the c++ code a bit and add a new class of functions, so I guess double check that what I did is valid on that front.

Next steps I am looking at are as follows

- expand the pkpy_push_* calls to support all the types we could want
- adjust the calling api so that it abstracts away the pkpy_push_none call from the user
- add the pkpy_to* for all the types we could want
- add pkpy_is_* calls to check for types
- work out the best way to manage errors so that exceptions within python don't cause a crash

I figured I would put up what I have so far, and hopefully get some feedback on if this is looking like what the project wants

**Comment by koltenpearson** [comment]
Okay with that push I anticipate that the basic functionality is looking pretty good. It now uses a separate stack for talking to c, and the calling api is looking much cleaner since we figured out a way to wrap the functions in a NativeFunc object. 

I also have adjusted the api to allow for error handling so that pkpy::Exeception won't just cause a crash and is error message is accessible.

Here are the next steps I am planning on taking to be able to merge 

- add in the pkpy_push_* calls
- add in the pkpy_to_* calls
- add in the pkpy_is_* calls
- add in a subset of the lua stack functions (at least getting the size and popping)
- correctly wrap and unwrap tuples for multiple return values
- make it so the c interop stack overflows and underflows throw an error instead of exceeding bounds silently
- write test cases

At that point the bindings will be fully usable and sufficient for a v1.0 api.

**Comment by koltenpearson** [comment]
Okay I would be ready to call these bindings done. Most all the bullet points from above are finished.

The only one not finished is adding bounds checks to the stack. While a little dangerous, it is pretty normal for c api to have to take care like that (lua's api doesn't do bounds checks on its stack). And there would be a speed penalty for doing so. 

The only changes that were made to the c++ code in the end were the creation of the LuaStyleFuncC, the creation of a new datastructure for the virtual c stack (a modified version of the ValueStack), and adding one of those datastructures as a member to the vm object.

As for organization, most of the new code is under the c_bindings folder. Trying to get the single header file lib working for c personally does not seem worth it to me, as you will need to compile it separately for c projects anyway to be able to link to it correctly. 

I guess the argument to roll it in would be because c++ wants to make use of the c bindings as well. If that is the case it can be done, but it will require changing how the amalgamation is compiled, and adding #ifdef __cplusplus in several places. Additionally the c program needs a header where it can only see the function declarations of the api, so you would need to separate the declaration and definition in any case. I might recommend making that work as a separate pull request.

File: src/frame.h
Comment (blueloveTH) [inline src/frame.h:1]: make a custom class elsewhere derived from `ValueStackImpl<256>` and discard the change in this file

File: c_bindings/pocketpy_c.h
Comment (blueloveTH) [inline c_bindings/pocketpy_c.h:1]: just use `VM*`, no need to make them opaque like this `typedef struct pkpy_vm_handle* pkpy_vm;`.

Comment (blueloveTH) [inline c_bindings/pocketpy_c.h:1]: make a custom class `LuaStyleVM` derived from `VM` and put virtual stack operations in the derived class

File: src/vm.h
Comment (blueloveTH) [inline src/vm.h:1]: 1. make a custom class `LuaStyleVM` derived from `VM` and put virtual stack operations in the derived class.
2. discard change in `src/vm.h`

**Comment by koltenpearson** [comment]
Okay I think I made the changes you requested. The stack size is now 32, as I have to be able to make a new c stack for each c bound function called, and 256 is probably a little big for that. 

I bet 32 will be workable though, it is not used like the call stack, so it does not grow with recursion or anything like that.

**Comment by koltenpearson** [comment]
updated with changes discussed on discord

- ironed out the strong and weak references returned by the various string methods
- switched to subclassing VM to make api cleaner
- added method to hook into the garbage collector

**Pull request merged**